protocol-specs/generators/multisignature/index.js

```diff
@@ -0,0 +1,221 @@
+		address: '7086965981385941478L',
+	},
+];
+
+const sortKeysDescending = publicKeys =>
+	publicKeys.sort((publicKeyA, publicKeyB) => {
+		if (publicKeyA < publicKeyB) return 1;
+		if (publicKeyA > publicKeyB) return -1;
+		return 0;
+	});
+
+const networkIdentifier =
+	'e48feb88db5b5cf5ad71d93cdcd1d879b6d5ed187a36b0002cc34e0ef9883255';
+
+const getId = transactionBytes => {
+	const transactionHash = hash(transactionBytes);
+	const bufferFromFirstEntriesReversed = getFirstEightBytesReversed(
+		transactionHash,
+	);
+	const transactionId = bufferToIntAsString(bufferFromFirstEntriesReversed);
+
+	return transactionId;
+};
+
+const assetToBytes = tx => {
+	const { mandatoryKeys, optionalKeys, numberOfSignatures } = tx.asset;
+	const mandatoryKeysBuffer = Buffer.from(mandatoryKeys.join(''), 'utf8');
+	const optionalKeysBuffer = Buffer.from(optionalKeys.join(''), 'utf8');
+	const assetBuffer = Buffer.concat([
+		intToBuffer(mandatoryKeys.length, 1),
+		mandatoryKeysBuffer,
+		intToBuffer(optionalKeys.length, 1),
+		optionalKeysBuffer,
+		intToBuffer(numberOfSignatures, 1),
+	]);
+
+	return assetBuffer;
+};
+
+const createSignatureObject = (txBuffer, account) => ({
+	// publicKey: account.publicKey,
+	signature: signData(
+		hash(Buffer.concat([hexToBuffer(networkIdentifier), txBuffer])),
+		account.passphrase,
+	),
+});
+
+const findAccountByKey = key =>
+	accounts.filter(account => account.publicKey === key);
+
+const generateValidMultisignatureRegistrationTransaction = () => {
+	// basic transaction
+	const tx = {
+		senderPublicKey: accounts[1].publicKey,
+		timestamp: 77045780,
+		type: 12,
+		asset: {
+			mandatoryKeys: [accounts[3].publicKey, accounts[2].publicKey],
+			optionalKeys: [accounts[4].publicKey, accounts[5].publicKey],
+			numberOfSignatures: 3,
+		},
+	};
+
+	sortKeysDescending(tx.asset.mandatoryKeys);
+	sortKeysDescending(tx.asset.optionalKeys);
+
+	const { signatures, ...registrationTx } = tx;
+	// Start building buffer
+	const transactionTimestamp = Buffer.alloc(4);
+	transactionTimestamp.writeIntBE(tx.timestamp, 0, 4);
+	let txBuffer = Buffer.concat([
+		Buffer.alloc(1, registrationTx.type),
+		transactionTimestamp,
+		hexToBuffer(tx.senderPublicKey),
+		intToBuffer(registrationTx.asset.numberOfSignatures, 1),
+		assetToBytes(registrationTx),
+	]);
+
+	registrationTx.signatures = [];
+	let index = 0;
+	// Mandatory keys sign
+	registrationTx.asset.mandatoryKeys.forEach(aKey => {
+		const account = findAccountByKey(aKey).pop();
+		registrationTx.signatures.push({
+			...createSignatureObject(txBuffer, account),
+			index,
+		});
+		index += 1;
+	});
+	// Optional keys sign
+	registrationTx.asset.optionalKeys.forEach(aKey => {
+		const account = findAccountByKey(aKey).pop();
+		registrationTx.signatures.push({
+			...createSignatureObject(txBuffer, account),
+			index,
+		});
+		index += 1;
+	});
+	// Serialize all signatures
+	registrationTx.signatures.forEach(aSignature => {
```

Comment:
instead of using `forEach`, i think it will be easier to use `map` or `reduce` to create the `signaturesBuffer`

Reply:
It's a good point. I'm going to try to refactor to be more expressive in general with the whole spec.

Reply:
I tried changing to both map and reduce but makes the code look more confusing. Reduce provides both previous and current values which we don't need together. Map makes the assignment of the index more complex as we need the index to be sequential for both mandatory and optional keys.